File: occasum-core/src/main/java/pw/stamina/occasum/node/visit/PrettyPrintingPropertyNodeVisitor.java
package pw.stamina.occasum.node.visit;

import pw.stamina.occasum.node.PropertyNode;

import java.io.PrintStream;
import java.util.Objects;
import java.util.stream.Stream;

public final class PrettyPrintingPropertyNodeVisitor extends AbstractPropertyNodeVisitor {
    private static final int DEFAULT_INDENTION = 0;
    private static final int INDENTION_PRINTING_OFFSET = 1;
    private static final int INDENTION_LEVEL_INCREASE = 1;

    private final PrintStream out;
    private final int indention;

    public PrettyPrintingPropertyNodeVisitor(PrintStream out) {
        this(out, DEFAULT_INDENTION);
    }

    private PrettyPrintingPropertyNodeVisitor(PrintStream out, int indention) {
        Objects.requireNonNull(out, "out");
        this.out = out;
        this.indention = indention;
    }

    @Override
    public void visitNode(PropertyNode node) {
        addIndention();

        out.print(node.getId());
        addPropertyValueIfPresent(node);

        out.println();
    }

    private void addIndention() {
        if (indention > 0) {
            int indentionExcludingFirstStep = indention - INDENTION_PRINTING_OFFSET;
            Stream.generate(() -> "  ").limit(indentionExcludingFirstStep).forEach(out::print);
            out.print("+ ");
        }
    }

    private void addPropertyValueIfPresent(PropertyNode node) {
        if (node.hasProperty()) {
            out.print(":");
            out.print(node.getProperty().getValueAsString());
        }
    }

    @Override
    protected PropertyNodeVisitor createChildVisitor(PropertyNode childNode) {
        return withIncreasedIndention();
    }

    private PrettyPrintingPropertyNodeVisitor withIncreasedIndention() {
        int increasedIndention = indention + INDENTION_LEVEL_INCREASE;
        return new PrettyPrintingPropertyNodeVisitor(out, increasedIndention);
    }
}

File: occasum-core/src/main/kotlin/pw/stamina/occasum/node/visit/PrettyPrintingPropertyNodeVisitor.kt
package pw.stamina.occasum.node.visit;

import pw.stamina.occasum.node.PropertyNode
import java.io.PrintStream
import java.util.stream.Stream

private const val DEFAULT_INDENTION = 0
private const val INDENTION_PRINTING_OFFSET = 1L
private const val INDENTION_LEVEL_INCREASE = 1

class PrettyPrintingPropertyNodeVisitor private constructor(
        private val out: PrintStream,
        private val indention: Int)
    : AbstractPropertyNodeVisitor() {

    constructor(out: PrintStream) : this(out, DEFAULT_INDENTION)

    override fun visitNode(node: PropertyNode) {
        addIndention()

        out.print(node.id)
        addPropertyValueIfPresent(node)

        out.println()
    }

    private fun addIndention() {
        if (indention > 0) {
            val indentionExcludingFirstStep: Long = indention - INDENTION_PRINTING_OFFSET


            Stream.generate({ "  " })
                    .limit(indentionExcludingFirstStep)
                    .forEach(out::print)

            out.print("+ ")
        }
    }

    private fun addPropertyValueIfPresent(node: PropertyNode) {
        val property = node.property//TODO ?: return

        out.print(":")
        out.print(property.valueAsString)
    }

    override fun createChildVisitor(childNode: PropertyNode): PropertyNodeVisitor {
        return withIncreasedIndention()
    }

    private fun withIncreasedIndention(): PrettyPrintingPropertyNodeVisitor {
        val increasedIndention = indention + INDENTION_LEVEL_INCREASE
        return PrettyPrintingPropertyNodeVisitor(out, increasedIndention)
    }
}
